### ISSUE TITLE
Rendering components in loop severely impacts performance

### ISSUE THREAD

**Issue by AntonCooper** [laravel/framework#43811]
- Laravel Version: 9.25.1
- PHP Version: 8.0.22
- Database Driver & Version: MySql 8.0.27

### Description:
Rendering components in loop severally impacts performance

To demonstrate this I have setup a simple test with two endpoints. They both render a list of 10000 book titles, generated from an array of strings. One endpoint uses a component to render each item in the list, the other uses a simple foreach loop on the data within the blade template.

`php artisan optimize` and `php artisan view:cache` have both been run.

Using XDebug's profiler to capture a snapshot of the requests to there two requests, I have observed the difference in time to render these is significant, with the component taking 20 times longer to render the same data; 
- Using component `4519ms`
- Using foreach `240ms`

In the call stack of the component version, a huge amount of total time is spent in;
`Illuminate\View\Component->resolveView()` - 1251ms
`Illuminate\View\Factory->renderComponent()` - 1421ms

While this test is an extreme example, we are seeing in our production system a significant impacts of this slow rendering of component in a loop - which is further compounded due to nested components. 

I think there probably needs to be some static caching introduced to help prevent the disk read of the component blade file on every loop, or a mention of this impact within the docs, to help guide people in when using a component is a good or bad idea.

### Steps To Reproduce:

See https://github.com/AntonCooper/laravel-component-performance for test repo with the two test endpoints.

See my profiler snapshots attached
[cachegrind.out.33270.gz](https://github.com/laravel/framework/files/9394732/cachegrind.out.33270.gz)
[cachegrind.out.33269.gz](https://github.com/laravel/framework/files/9394733/cachegrind.out.33269.gz)

**Comment by driesvints** [comment]
Hi there. Using loops to render components can indeed be performance intensive in some situations like your own. But view caching should greatly mitigate this. In general, only the first rendering should be an expensive operation. If you still have performance issues I suggest to look at other solutions like https://github.com/spatie/laravel-responsecache. Thanks

**Comment by AntonCooper** [comment]
HI @driesvints, thanks for getting back to me so quickly. 
Unfortunately the view caching does nothing to mitigate the slowness of the test I've outlined above. I can see from the stack trace that it is using the compiled blade template - but the slowness persists. It seems to be reading that rendered blade template from disk on every single loop. This seems to be the root cause of the slowness. 

The test is extremely simplistic to try and demonstrate the internal overhead of trying to load components in a loop. I understand using components might have an additional overhead, but 20 times slower seems like pretty significant difference which at worst seems to be a performance issue. If it's working as designed I really feel that these performance overheads need to be stated in the docs as it will greatly influence the way which they're used within a project.

**Comment by GC-Mark** [comment]
Just to add some more info to this, we saw a similar issue when using https://github.com/blade-ui-kit/blade-icons when using components.

Discussion and benchmarks here - https://github.com/blade-ui-kit/blade-icons/issues/190

**Comment by driesvints** [comment]
Seems there is no good way of improving this. In these use cases it's probably best to use something differently than components.

**Comment by sha-hin** [comment]
Same problem here.
I have always not used components too much except in few cases, but since I've decided to use Tailwind in my new project, I thought it would be better to make components for different pieces on the website then use those components without rewriting the classes and styles everywhere again and again, but I've noticed that the performance is very bad even I'm still building the front-end, no DB queries or business logic has been written yet.
So nothing to blame except the components, then after I saw how blade compiled the views with the components, then I knew why the performance dropped to this level.
Of course I've removed most of already created components and have to copy and paste the repetitive HTML code again and again, yes bad thing to do but performance wise waaaaaay better.
I think we need a way to flatten the views not to create component instances and render them inside the compiled version, it makes no sense.

**Comment by JayBizzle** [comment]
See #44480

**Comment by johnbacon** [comment]
Three years later and this is unfortunately still the case.[^1]

If there's nothing that can be done about this, that would be great to know. We are having to rely on the include directive and other techniques that others in the community have long recognized as they work around these problems: https://laravel-news.com/faster-laravel-optimizations

This limits the usefulness of Blade components significantly for an e-commerce/marketplace like ours, with results cards containing multiple components -- a common thing on the web.

tl;dr: Laravel 12.x components are still multiple times slower than the include directive and exponentially slower than reusable PHP partials.

[^1]: One clarification: it's not caused by components within loops, but instead by using "too many" components. Just so happens that components within loops often equate to many components.
